File: app/src/main/java/com/cool/eye/scan/view/CaptureView.java
package com.cool.eye.scan.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Paint.Style;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.View;
import androidx.annotation.DrawableRes;
import com.cool.eye.scan.R;
import com.google.zxing.ResultPoint;
import java.util.LinkedList;

public class CaptureView extends View {

  private class PossiblePoint {

    public float x, y;
    public long foundTime;
  }

  public static final int MASK_COLOR = 0x80000000;
  public static final int POSSIBLE_POINT_COLOR = 0xC0FFFF00;
  public static final int POSSIBLE_POINT_ALIVE_DURATION = 200;
  public static final int SCAN_DURATION = 1500;

  private int maskColor = MASK_COLOR;
  private int possiblePointColor = POSSIBLE_POINT_COLOR;
  private int possiblePointAliveDuration = POSSIBLE_POINT_ALIVE_DURATION;
  private int scanDuration = SCAN_DURATION;
  private long startTime = -1;

  private Rect frame;
  private Paint paint;
  private LinkedList<PossiblePoint> possiblePoints;
  private Drawable frameDrawable, scannerDrawable;
  private int scannerHeight = 0;

  public CaptureView(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
    init(context, attrs);
  }

  public CaptureView(Context context, AttributeSet attrs) {
    super(context, attrs);
    init(context, attrs);
  }

  public CaptureView(Context context) {
    super(context);
    init(context, null);
  }

  public void init(Context context, AttributeSet attrs) {
    frame = new Rect();
    paint = new Paint();
    paint.setAntiAlias(true);
    possiblePoints = new LinkedList<>();
    int frameResId = R.drawable.scan_qrcode_frame;
    int scannerResId = R.drawable.scan_qrcode_scaner;
    if (attrs != null) {
      TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.scan_captureView);
      maskColor = array.getColor(R.styleable.scan_captureView_scan_mask_color, MASK_COLOR);
      possiblePointColor = array
          .getColor(R.styleable.scan_captureView_scan_possible_point_color, POSSIBLE_POINT_COLOR);
      possiblePointAliveDuration = array
          .getInt(R.styleable.scan_captureView_scan_possible_point_alive_duration,
              POSSIBLE_POINT_ALIVE_DURATION);
      scanDuration = array.getInt(R.styleable.scan_captureView_scan_duration, SCAN_DURATION);
      frameResId = array
          .getResourceId(R.styleable.scan_captureView_scan_frame_drawable, R.drawable.scan_qrcode_frame);
      scannerResId = array
          .getResourceId(R.styleable.scan_captureView_scan_scanner_drawable, R.drawable.scan_qrcode_scaner);
      array.recycle();
    }

    frameDrawable = getResources().getDrawable(frameResId);
    scannerDrawable = getResources().getDrawable(scannerResId);
  }

  public void setMaskColor(int maskColor) {
    this.maskColor = maskColor;
    invalidate();
  }

  public void setPossiblePointAliveDuration(int possiblePointAliveDuration) {
    this.possiblePointAliveDuration = possiblePointAliveDuration;
    invalidate();
  }

  public void setPossiblePointColor(int color) {
    this.possiblePointColor = color;
    invalidate();
  }

  public void setScanDuration(int duration) {
    scanDuration = duration;
    invalidate();
  }

  public void setFrameDrawableRes(@DrawableRes int resId) {
    frameDrawable = getResources().getDrawable(resId);
    invalidate();
  }

  public void setScannerDrawableRes(@DrawableRes int resId) {
    scannerDrawable = getResources().getDrawable(resId);
    invalidate();
  }

  @Override
  protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
    super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    int width = getMeasuredWidth();
    int height = getMeasuredHeight();
    int length = (int) (width * 0.6);
    frame.left = width / 2 - length / 2;
    frame.right = width / 2 + length / 2;
    frame.top = height / 2 - length / 2;
    frame.bottom = height / 2 + length / 2;
    frameDrawable.setBounds(frame.left - 10, frame.top - 10, frame.right + 10, frame.bottom + 10);
    scannerHeight =
        scannerDrawable.getIntrinsicHeight() * frame.width() / scannerDrawable.getIntrinsicWidth();
  }

  public Rect getFrameRect() {
    return frame;
  }

  @Override
  protected void onDraw(Canvas canvas) {

    // Draw mask
    paint.setColor(maskColor);
    paint.setStyle(Style.FILL);
    int width = canvas.getWidth();
    int height = canvas.getHeight();
    canvas.drawRect(0, 0, width, frame.top, paint);
    canvas.drawRect(0, frame.top, frame.left, frame.bottom, paint);
    canvas.drawRect(frame.right, frame.top, width, frame.bottom, paint);
    canvas.drawRect(0, frame.bottom, width, height, paint);

    // Draw possible points
    paint.setColor(possiblePointColor);
    paint.setStyle(Style.FILL);
    long current = System.currentTimeMillis();
    while (possiblePoints.size() > 0
        && current - possiblePoints.peek().foundTime >= possiblePointAliveDuration) {
      possiblePoints.poll();
    }
    for (int i = 0; i < possiblePoints.size(); i++) {
      PossiblePoint point = possiblePoints.get(i);
      int radius = (int) (5 * (possiblePointAliveDuration - current + point.foundTime)
          / possiblePointAliveDuration);
      if (radius > 0) {
        canvas.drawCircle(frame.left + point.x, frame.top + point.y, radius, paint);
      }
    }

    // Draw scanner
    long now = System.currentTimeMillis();
    if (startTime < 0) {
      startTime = now;
    }
    int timePast = (int) ((now - startTime) % scanDuration);
    if (timePast >= 0 && timePast <= scanDuration / 2) {
      int scannerShift = frame.height() * 2 * timePast / scanDuration;
      canvas.save();
      canvas.clipRect(frame);
      scannerDrawable.setBounds(frame.left, frame.top + scannerShift, frame.right,
          frame.top + scannerHeight + scannerShift);
      scannerDrawable.draw(canvas);
      canvas.restore();
    }
    // Draw frame
    frameDrawable.draw(canvas);

    invalidate();
  }

  public void addPossibleResultPoint(ResultPoint point) {
    PossiblePoint pp = new PossiblePoint();
    pp.foundTime = System.currentTimeMillis();
    pp.x = point.getX();
    pp.y = point.getY();
    if (possiblePoints.size() >= 10) {
      possiblePoints.poll();
    }
    possiblePoints.add(pp);
  }
}

File: app/src/main/java/com/eye/cool/scan/decode/view/CaptureView.kt
package com.eye.cool.scan.decode.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import androidx.core.content.ContextCompat
import com.eye.cool.scan.R
import com.google.zxing.ResultPoint
import java.util.*

class CaptureView : View {

  private inner class PossiblePoint {
    var x = 0f
    var y = 0f
    var foundTime: Long = 0
  }

  val frameRect = Rect()
  private var startTime: Long = -1
  private val paint = Paint()
  private var possiblePoints = LinkedList<PossiblePoint>()
  private var scannerHeight = 0
  private var descX = 0f
  private var descY = 0f

  private var params = CaptureParams.Builder().build()
    set(value) {
      field = value
      invalidate()
    }

  private var frameDrawable: Drawable? = null
  private var scannerDrawable: Drawable? = null

  constructor(context: Context) : super(context) {
    init(context, null)
  }

  constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
    init(context, attrs)
  }

  private fun init(context: Context, attrs: AttributeSet?) {
    paint.isAntiAlias = true
    paint.isDither = true
    attrs?.let {
      val array = context.obtainStyledAttributes(it, R.styleable.scan_capture_view)
      params.maskColor = array.getColor(R.styleable.scan_capture_view_scan_mask_color, MASK_COLOR)
      params.possiblePointColor = array.getColor(
          R.styleable.scan_capture_view_scan_possible_point_color,
          POSSIBLE_POINT_COLOR
      )
      params.possiblePointAliveDuration = array.getInt(
          R.styleable.scan_capture_view_scan_possible_point_alive_duration,
          POSSIBLE_POINT_ALIVE_DURATION
      )
      params.scanDuration = array.getInt(R.styleable.scan_capture_view_scan_duration, SCAN_DURATION)
      params.frameResId = array.getResourceId(
          R.styleable.scan_capture_view_scan_frame_drawable,
          R.drawable.scan_frame_shape
      )
      params.scannerResId = array.getResourceId(
          R.styleable.scan_capture_view_scan_scanner_drawable,
          R.drawable.scan_scanner_shape
      )
      params.frameWidthRatio = array.getFloat(
          R.styleable.scan_capture_view_scan_frame_width_ratio,
          0.6f
      )
      params.frameHeightRatio = array.getFloat(
          R.styleable.scan_capture_view_scan_frame_height_ratio,
          0.4f
      )
      params.scanDesc = array.getString(R.styleable.scan_capture_view_scan_description)
      params.descTextColor = array.getColor(
          R.styleable.scan_capture_view_scan_text_color,
          Color.WHITE
      )
      params.descTextSize = array.getDimension(
          R.styleable.scan_capture_view_scan_text_size,
          context.resources.displayMetrics.density * 15f
      )
      params.descOffsetXRatio = array.getFloat(
          R.styleable.scan_capture_view_scan_text_offset_x_ratio,
          0f
      )
      params.descOffsetYRatio = array.getFloat(
          R.styleable.scan_capture_view_scan_text_offset_y_ratio,
          params.frameHeightRatio / 2.0f + 0.03f
      )
      array.recycle()
    }
    frameDrawable = ContextCompat.getDrawable(context, params.frameResId)
    scannerDrawable = ContextCompat.getDrawable(context, params.scannerResId)
  }


  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
    super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    val width = measuredWidth
    val height = measuredHeight
    val frameWidth = (width * params.frameWidthRatio).toInt()
    val frameHeight = (height * params.frameHeightRatio).toInt()
    frameRect.left = (width - frameWidth) / 2
    frameRect.right = (width + frameWidth) / 2
    frameRect.top = (height - frameHeight) / 2
    frameRect.bottom = (height + frameHeight) / 2
    val offset = params.frameOffset
    frameDrawable?.setBounds(
        frameRect.left - offset,
        frameRect.top - offset,
        frameRect.right + offset,
        frameRect.bottom + offset
    )
    scannerDrawable?.apply {
      scannerHeight = intrinsicHeight * frameRect.width() / intrinsicWidth
    }

    params.scanDesc?.apply {
      paint.textSize = params.descTextSize
      descX = (width - paint.measureText(this)) / 2f + params.descOffsetXRatio * width
      descY = height / 2 + paint.descent() - paint.ascent() + params.descOffsetYRatio * height
    }
  }


  override fun onDraw(canvas: Canvas) {
    drawMask(canvas)
    drawPoints(canvas)
    drawDrawable(canvas)
    drawDesc(canvas)

    invalidate()
  }

  private fun drawMask(canvas: Canvas) {
    paint.color = params.maskColor
    paint.style = Paint.Style.FILL
    val width = canvas.width
    val height = canvas.height
    canvas.drawRect(0f, 0f, width.toFloat(), frameRect.top.toFloat(), paint)
    canvas.drawRect(
        0f,
        frameRect.top.toFloat(),
        frameRect.left.toFloat(),
        frameRect.bottom.toFloat(), paint
    )
    canvas.drawRect(
        frameRect.right.toFloat(),
        frameRect.top.toFloat(),
        width.toFloat(),
        frameRect.bottom.toFloat(), paint
    )
    canvas.drawRect(0f, frameRect.bottom.toFloat(), width.toFloat(), height.toFloat(), paint)
  }

  private fun drawPoints(canvas: Canvas) {
    if (possiblePoints.isEmpty()) return
    paint.color = params.possiblePointColor
    paint.style = Paint.Style.FILL
    val current = System.currentTimeMillis()
    val duration = params.possiblePointAliveDuration
    while (current - (possiblePoints.peek()?.foundTime ?: current) >= duration) {
      possiblePoints.poll()
    }
    if (possiblePoints.isEmpty()) return
    for (i in possiblePoints.indices) {
      val point = possiblePoints[i]
      val radius = 5f * (duration - current + point.foundTime) / duration
      if (radius > 0f) {
        canvas.drawCircle(frameRect.left + point.x, frameRect.top + point.y, radius, paint)
      }
    }
  }

  private fun drawDrawable(canvas: Canvas) {
    scannerDrawable?.apply {
      val now = System.currentTimeMillis()
      if (startTime < 0) {
        startTime = now
      }
      val timePast = ((now - startTime) % params.scanDuration).toInt()
      if (timePast > 0 && timePast <= params.scanDuration / 2) {
        val scannerShift = frameRect.height() * 2 * timePast / params.scanDuration
        canvas.save()
        canvas.clipRect(frameRect!!)
        setBounds(frameRect.left, frameRect.top + scannerShift, frameRect.right,
            frameRect.top + scannerHeight + scannerShift)
        draw(canvas)
        canvas.restore()
      }
    }

    frameDrawable?.draw(canvas)
  }

  private fun drawDesc(canvas: Canvas) {
    val desc = params.scanDesc ?: return
    paint.color = params.descTextColor
    paint.textSize = params.descTextSize
    canvas.drawText(desc, descX, descY, paint)
  }

  internal fun addPossibleResultPoint(point: ResultPoint) {
    val pp = PossiblePoint()
    pp.foundTime = System.currentTimeMillis()
    pp.x = point.x
    pp.y = point.y
    if (possiblePoints.size >= 10) {
      possiblePoints.poll()
    }
    possiblePoints.add(pp)
  }

  internal companion object {
    const val MASK_COLOR = -0x80000000
    const val POSSIBLE_POINT_COLOR = -0x3f000100
    const val POSSIBLE_POINT_ALIVE_DURATION = 200
    const val SCAN_DURATION = 1500
  }
}